### ISSUE TITLE
'Repository' object has no attribute 'segments'

### ISSUE THREAD

**Issue by skibbipl** [borgbackup/borg#502]
So I upgraded server and client to borg 0.29 and one of my backups stopped working.
`borg create --compression=lzma,9 --stats -v --filter=AME` generates following errors:
```
Remote: Borg 0.29.0: exception in RPC call:
Remote: Traceback (most recent call last):
Remote:   File "/usr/lib/python3/dist-packages/borg/remote.py", line 96, in serve
Remote:     res = f(*args)
Remote:   File "/usr/lib/python3/dist-packages/borg/repository.py", line 439, in put
Remote:     self.segments.setdefault(segment, 0)
Remote: AttributeError: 'Repository' object has no attribute 'segments'
Remote: Platform: Linux copam 4.3.0-1-amd64 #1 SMP Debian 4.3.3-1 (2015-12-15) x86_64
Remote: Linux: debian stretch/sid   LibC: glibc 2.9
Remote: Python: CPython 3.4.4rc1
```
Both client and remote are debian servers.

UPDATE: OK, It seems I need some sleep. I forgot that I mounted my backup on remote.

**Comment by ThomasWaldmann** [comment]
well, it maybe should try to handle that more gracefully, reopening.

**Comment by ThomasWaldmann** [comment]
I could repoduce (mount a repo locally, then access same repo via ssh://localhost/...), full traceback:
```
Remote: Traceback (most recent call last):
Remote:   File "/home/tw/w/borg/borg/remote.py", line 96, in serve
Remote:     res = f(*args)
Remote:   File "/home/tw/w/borg/borg/repository.py", line 428, in put
Remote:     self.prepare_txn(self.get_transaction_id())
Remote:   File "/home/tw/w/borg/borg/repository.py", line 177, in prepare_txn
Remote:     self.lock.upgrade()
Remote:   File "/home/tw/w/borg/borg/locking.py", line 303, in upgrade
Remote:     self.acquire(exclusive=True, remove=SHARED)
Remote:   File "/home/tw/w/borg/borg/locking.py", line 264, in acquire
Remote:     self._wait_for_readers_finishing(remove, sleep)
Remote:   File "/home/tw/w/borg/borg/locking.py", line 291, in _wait_for_readers_finishing
Remote:     raise LockTimeout(self.path)
Remote: borg.locking.LockTimeout: /home/tw/w/borg/repo2/lock

Remote: Borg 0.29.1.dev7+ng546052e: exception in RPC call:
Remote: Traceback (most recent call last):
Remote:   File "/home/tw/w/borg/borg/remote.py", line 96, in serve
Remote:     res = f(*args)
Remote:   File "/home/tw/w/borg/borg/repository.py", line 428, in put
Remote:     self.prepare_txn(self.get_transaction_id())
Remote:   File "/home/tw/w/borg/borg/repository.py", line 177, in prepare_txn
Remote:     self.lock.upgrade()
Remote:   File "/home/tw/w/borg/borg/locking.py", line 303, in upgrade
Remote:     self.acquire(exclusive=True, remove=SHARED)
Remote:   File "/home/tw/w/borg/borg/locking.py", line 264, in acquire
Remote:     self._wait_for_readers_finishing(remove, sleep)
Remote:   File "/home/tw/w/borg/borg/locking.py", line 280, in _wait_for_readers_finishing
Remote:     self._roster.modify(remove, REMOVE)
Remote:   File "/home/tw/w/borg/borg/locking.py", line 217, in modify
Remote:     elements.remove(self.id)
Remote: KeyError: ('tux', 22968, 0)

Remote: Borg 0.29.1.dev7+ng546052e: exception in RPC call:
Remote: Traceback (most recent call last):
Remote:   File "/home/tw/w/borg/borg/remote.py", line 96, in serve
Remote:     res = f(*args)
Remote:   File "/home/tw/w/borg/borg/repository.py", line 428, in put
Remote:     self.prepare_txn(self.get_transaction_id())
Remote:   File "/home/tw/w/borg/borg/repository.py", line 177, in prepare_txn
Remote:     self.lock.upgrade()
Remote:   File "/home/tw/w/borg/borg/locking.py", line 303, in upgrade
Remote:     self.acquire(exclusive=True, remove=SHARED)
Remote:   File "/home/tw/w/borg/borg/locking.py", line 264, in acquire
Remote:     self._wait_for_readers_finishing(remove, sleep)
Remote:   File "/home/tw/w/borg/borg/locking.py", line 280, in _wait_for_readers_finishing
Remote:     self._roster.modify(remove, REMOVE)
Remote:   File "/home/tw/w/borg/borg/locking.py", line 217, in modify
Remote:     elements.remove(self.id)
Remote: KeyError: ('tux', 22968, 0)

Remote: Borg 0.29.1.dev7+ng546052e: exception in RPC call:
Remote: Traceback (most recent call last):
Remote:   File "/home/tw/w/borg/borg/remote.py", line 96, in serve
Remote:     res = f(*args)
Remote:   File "/home/tw/w/borg/borg/repository.py", line 439, in put
Remote:     self.segments.setdefault(segment, 0)
Remote: AttributeError: 'Repository' object has no attribute 'segments'

Remote: Borg 0.29.1.dev7+ng546052e: exception in RPC call:
Remote: Traceback (most recent call last):
Remote:   File "/home/tw/w/borg/borg/remote.py", line 96, in serve
Remote:     res = f(*args)
Remote:   File "/home/tw/w/borg/borg/repository.py", line 439, in put
Remote:     self.segments.setdefault(segment, 0)
Remote: AttributeError: 'Repository' object has no attribute 'segments'

(... and many more same as last one ...)
```

**Comment by ThomasWaldmann** [comment]
This is:
a) a locking issue (if it can't get the exclusive lock [see first rpc call / first exception], it should not behave otherwise [see 2nd call / 2nd exception])
b) an exception handling issue: if it gets a LockTimeout exception from the server, the client should abort immediately.

**Comment by ThomasWaldmann** [comment]
Part b) was filed as #562, so this ticket will be restricted to part a) (== the root cause for what the OP has seen) now.

**Comment by ThomasWaldmann** [comment]
Fixed by #561.

**Comment by ThomasWaldmann** [comment]
Part b) fixed by same fix as #1220.
